plugins/SpectrumAnalyzer/SaWaterfallView.cpp

```diff
@@ -120,44 +139,70 @@ void SaWaterfallView::paintEvent(QPaintEvent *event)
 	// draw the spectrogram precomputed in SaProcessor
 	if (m_processor->m_waterfallNotEmpty)
 	{
-		QMutexLocker lock(&m_processor->m_dataAccess);
-		painter.drawImage(displayLeft, displayTop,					// top left corner coordinates
-						  QImage(m_processor->m_history.data(),		// raw pixel data to display
-								 m_processor->binCount(),			// width = number of frequency bins
-								 m_processor->m_waterfallHeight,	// height = number of history lines
-								 QImage::Format_RGB32
-								 ).scaled(displayWidth,				// scale to fit view..
-										  displayBottom,
-										  Qt::IgnoreAspectRatio,
-										  Qt::SmoothTransformation));
+		QMutexLocker lock(&m_processor->m_reallocationAccess);
+		QImage temp = QImage(m_processor->m_history.data(),		// raw pixel data to display
+							 m_processor->waterfallWidth(),		// width = number of frequency bins
+							 m_processor->m_waterfallHeight,	// height = number of history lines
+							 QImage::Format_RGB32);
 		lock.unlock();
+		temp.setDevicePixelRatio(devicePixelRatio());			// display at native resolution
+		painter.drawImage(m_displayLeft, m_displayTop,
+						  temp.scaled(m_displayWidth * devicePixelRatio(),
+									  m_displayHeight * devicePixelRatio(),
+									  Qt::IgnoreAspectRatio,
+									  Qt::SmoothTransformation));
+		m_processor->flipRequest();
 	}
 	else
 	{
-		painter.fillRect(displayLeft, displayTop, displayWidth, displayBottom, QColor(0,0,0));
+		painter.fillRect(m_displayLeft, m_displayTop, m_displayWidth, m_displayHeight, QColor(0,0,0));
 	}
 
+	// draw cursor (if it is within bounds)
+	drawCursor(painter);
+
 	// always draw the outline
 	painter.setPen(QPen(m_controls->m_colorGrid, 2, Qt::SolidLine, Qt::RoundCap, Qt::BevelJoin));
-	painter.drawRoundedRect(displayLeft, displayTop, displayWidth, displayBottom, 2.0, 2.0);
+	painter.drawRoundedRect(m_displayLeft, m_displayTop, m_displayWidth, m_displayHeight, 2.0, 2.0);
 
 	#ifdef SA_DEBUG
-		// display what FPS would be achieved if waterfall ran in a loop
-		start_time = std::chrono::high_resolution_clock::now().time_since_epoch().count() - start_time;
+		draw_time = std::chrono::high_resolution_clock::now().time_since_epoch().count() - draw_time;
+		m_execution_avg = 0.95 * m_execution_avg + 0.05 * draw_time / 1000000.0;
 		painter.setPen(QPen(m_controls->m_colorLabels, 1, Qt::SolidLine, Qt::RoundCap, Qt::BevelJoin));
-		painter.drawText(displayRight -100, 10, 100, 16, Qt::AlignLeft,
-						 QString(std::string("Max FPS: " + std::to_string(1000000000.0 / start_time)).c_str()));
+		painter.drawText(m_displayRight -150, 10, 100, 16, Qt::AlignLeft,
+						 QString(std::string("Exec avg.: " + std::to_string(m_execution_avg).substr(0, 5) + " ms").c_str()));
 	#endif
 }
 
 
+// Helper functions for time conversion
+float SaWaterfallView::samplesPerLine()
+{
+	return (float)m_processor->m_inBlockSize / m_controls->m_windowOverlapModel.value();
+}
+
+float SaWaterfallView::secondsPerLine()
```

Comment:
Normally a plugin should probably use `sample_t` (which is currently `float` btw.) so that if LMMS ever decides `double` precision is the way to go, all of the code does not have to be changed. I explicitly used `float` everywhere because the spectrum analyzer is quite performance hungry as it is and I highly doubt the rounding errors will have any noticeable impact on the results. AFAIK, some GPUs (or all of them) even support `half` precision for even faster and less precise float operations, because in graphics you usually can get away with it.

Reply:
LOL no I meant the opposite :) I might have used 8some kind of) integers, to avoid any kind of rounding errors. But I'm not sure if it makes sense here.

Reply:
Ah.. :)) In this case it is probably not that critical, the maximum error should be only around 0.5 %, but it would show up in measurements done using the cursor, so I prefer having at least _some_ accuracy in there in case anything changes and makes the error bigger or easier to see.